### PR TITLE
[RISCV] Handle fixed length vectors with exact VLEN in lowerINSERT_SUBVECTOR

### PR DESCRIPTION
**Pull request by lukel97**
This is the insert_subvector equivalent to #79949, where we can avoid sliding up by the full LMUL amount if we know the exact subregister the subvector will be inserted into.

This mirrors the lowerEXTRACT_SUBVECTOR changes in that we handle this in two parts:

- We handle fixed length subvector types by converting the subvector to a scalable vector. But unlike EXTRACT_SUBVECTOR, we may also need to convert the vector being inserted into too.

- Whenever we don't need a vslideup because either the subvector fits exactly into a vector register group *or* the vector is undef, we need to emit an insert_subreg ourselves because RISCVISelDAGToDAG::Select doesn't correctly handle fixed length subvectors yet: see d7a28f7ad

A subvector exactly fits into a vector register group if its size is a known multiple of the size of a vector register, and this adds a new overload for TypeSize::isKnownMultipleOf for scalable to scalable comparisons to help reason about this.

I've left RISCVISelDAGToDAG::Select untouched for now (minus relaxing an invariant), so that the insert_subvector and extract_subvector code paths are the same.

We should teach it to properly handle fixed length subvectors in a follow-up patch, so that the "exact subregsiter" logic is handled in one place instead of being spread across both RISCVISelDAGToDAG.cpp and RISCVISelLowering.cpp.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 1]
```diff
@@ -2063,8 +2063,14 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
     MVT SubVecContainerVT = SubVecVT;
     // Establish the correct scalable-vector types for any fixed-length type.
     if (SubVecVT.isFixedLengthVector()) {
-      assert(Idx == 0 && V.isUndef());
       SubVecContainerVT = TLI.getContainerForFixedLengthVector(SubVecVT);
+      bool AlignedToVecReg = false;
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:2067]: Two points:
1) Naming - This doesn't appear to be checking alignment, it appears to be checking whether the fixed vector completely fills the container.
2) Variable used under assert only should be in appropriate ifndef block.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -9596,6 +9596,21 @@ SDValue RISCVTargetLowering::lowerVPREDUCE(SDValue Op,
...
 
+/// Returns true if \p LHS is known to be equal to \p RHS, taking into account
+/// if VLEN is exactly known by \p Subtarget and thus vscale when handling
+/// scalable quantities.
+static bool isKnownEQ(ElementCount LHS, ElementCount RHS,
+                      const RISCVSubtarget &Subtarget) {
+  if (auto VLen = Subtarget.getRealVLen()) {
+    const unsigned Vscale = *VLen / RISCV::RVVBitsPerBlock;
+    if (LHS.isScalable())
+      LHS = ElementCount::getFixed(LHS.getKnownMinValue() * Vscale);
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:9607]: You've got this pattern repeating a few times, maybe extract a helper which returns the ElementCount of a vector which returns a fixed value if possible given exact vlen?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -9698,41 +9714,92 @@ SDValue RISCVTargetLowering::lowerINSERT_SUBVECTOR(SDValue Op,
...
+            ContainerVecVT, ContainerSubVecVT, OrigIdx, TRI);
+    SubRegIdx = Decompose.first;
+    RemIdx = ElementCount::getScalable(Decompose.second);
+  }
+
+  RISCVII::VLMUL SubVecLMUL = RISCVTargetLowering::getLMUL(ContainerSubVecVT);
   bool IsSubVecPartReg = SubVecLMUL == RISCVII::VLMUL::LMUL_F2 ||
                          SubVecLMUL == RISCVII::VLMUL::LMUL_F4 ||
                          SubVecLMUL == RISCVII::VLMUL::LMUL_F8;
+  bool AlignedToVecReg = !IsSubVecPartReg;
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:9755]: Again, not alignment in the naming here, and another instance of getting a fixed size from a scalable quantity.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -9698,41 +9714,92 @@ SDValue RISCVTargetLowering::lowerINSERT_SUBVECTOR(SDValue Op,
...
   // 1. If the Idx has been completely eliminated and this subvector's size is
   // a vector register or a multiple thereof, or the surrounding elements are
   // undef, then this is a subvector insert which naturally aligns to a vector
   // register. These can easily be handled using subregister manipulation.
-  // 2. If the subvector is smaller than a vector register, then the insertion
-  // must preserve the undisturbed elements of the register. We do this by
-  // lowering to an EXTRACT_SUBVECTOR grabbing the nearest LMUL=1 vector type
-  // (which resolves to a subregister copy), performing a VSLIDEUP to place the
-  // subvector within the vector register, and an INSERT_SUBVECTOR of that
+  // 2. If the subvector isn't exactly aligned to a vector register group, then
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:9765]: I think you actually want something along lines of "is an exact multiple of a valid register group size".

**Review by preames — COMMENTED** [review]
This looks reasonable likely to be correct.  I've added a couple small comments on style to make things easier to follow, but I don't see any major problems.

Note to future self - I've glanced at code only.  Need to go through the test diffs carefully to see if that raises any questions.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 3 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: I've gone a step further here to try and simplify this. We can check if the SubVec is known to completely fill the container if its size is a multiple of a vector register.

We already have a `isKnownMultipleOf` method on TypeSize that works on scalar quantities, but I've added a new variant that also works on scalable quantities. 

This way we can unify the fixed and scalable reasoning.

**Comment by lukel97** [comment]
Ping

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 5]
```diff
@@ -2063,8 +2063,16 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
     MVT SubVecContainerVT = SubVecVT;
     // Establish the correct scalable-vector types for any fixed-length type.
     if (SubVecVT.isFixedLengthVector()) {
-      assert(Idx == 0 && V.isUndef());
       SubVecContainerVT = TLI.getContainerForFixedLengthVector(SubVecVT);
+#ifndef NDEBUG
+      TypeSize VecRegSize = TypeSize::getScalable(RISCV::RVVBitsPerBlock);
+      bool ExactlyVecRegSized =
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp]: Doesn't the `#ifndef NDEBUG` make the [[maybe_unused]] unnecessary?

**Review by preames — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp]: We probably want to actually run these assertions on release builds with assertions enabled, so I'll remove the NDEBUG check

**Pull request merged**